File: ugmapid.py
#!/usr/bin/env python3

import argparse
import os
import stat


# walk_files performs the actual uid/gid change
def walk_files(root: str, reference, base: int, verbose: bool = True) -> None:
    for basedir, dirs, files in os.walk(root):
        files_map = map(lambda f: os.path.join(basedir, f), files)
        directories_map = map(lambda d: os.path.join(basedir, d), dirs)
        for path in list(files_map) + list(directories_map) + [root]:
            st = os.lstat(path)
            tgt_uid, tgt_gid = st.st_uid, st.st_gid
            if tgt_uid < base or tgt_uid > base + 65536:
                tgt_uid = base + (tgt_uid - reference)
            if tgt_gid < base or tgt_gid > base + 65536:
                tgt_gid = base + (tgt_gid - reference)
            if tgt_uid != st.st_uid or tgt_gid != st.st_gid:
                os.lchown(path, tgt_uid, tgt_gid)
                if stat.S_ISDIR(st.st_mode) or stat.S_ISREG(st.st_mode):
                    os.chmod(path, st.st_mode)
                if verbose:
                    print('{0}: uid {1} -> {2}, gid {3} -> {4}'.format(path, st.st_uid, tgt_uid, st.st_gid, tgt_gid))


# main program
if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--path', type=str, required=True, help='Root directory to process')
    parser.add_argument('--base', type=int, default=100000, help='Uid/Gid base offset')
    parser.add_argument('--verbose', action='store_true', help='Verbose output')
    args = parser.parse_args()

    # update base
    base_id = os.stat(args.path).st_uid
    if args.verbose:
        print('Calculated base offset: {0}'.format(base_id))
    walk_files(args.path, base_id, args.base, args.verbose)

File: ugmapid.go
package main

import (
	"flag"
	"log"
	"os"
	"path/filepath"
	"syscall"
)

/* change uid and gid for a single path element */
func walkFn(base, offset int, verbose bool) filepath.WalkFunc {
	return func(path string, info os.FileInfo, err error) error {
		stat := info.Sys().(*syscall.Stat_t)
		/* get target uid and gid values */
		tgt_uid, tgt_gid := int(stat.Uid), int(stat.Gid)
		/* check ranges */
		if tgt_uid < offset || tgt_uid > offset+65536 {
			tgt_uid = offset + (tgt_uid - base)
		}
		if tgt_gid < offset || tgt_gid > offset+65536 {
			tgt_gid = offset + (tgt_gid - base)
		}
		/* change file/directory owner */
		if tgt_uid != int(stat.Uid) || tgt_gid != int(stat.Gid) {
			if err = os.Lchown(path, tgt_uid, tgt_gid); err != nil {
				log.Println("  ERROR: ", path)
				return err
			}
			/* show some progress */
			if verbose {
				log.Printf("%s : uid %d -> %d, gid %d -> %d\n",
					path, stat.Uid, tgt_uid, stat.Gid, tgt_gid)
			}
		}
		return nil
	}
}

/* main program */
func main() {
	/* initialize arguments */
	var (
		directory string
		offset    int
		verbose   bool
	)
	flag.StringVar(
		&directory,
		"directory",
		"",
		"Filesystem directory to traverse")
	flag.IntVar(
		&offset,
		"offset",
		100000,
		"uid/gid map offset")
	flag.BoolVar(
		&verbose,
		"verbose",
		false,
		"Display verbose output")
	flag.Parse()
	/* check proper directory value */
	if len(directory) == 0 {
		log.Fatal("error: directory argument is required")
	}
	st, err := os.Stat(directory)
	if err != nil {
		log.Fatal("error:", err)
	}
	if !st.Mode().IsDir() {
		log.Fatal("error: not a directory:", directory)
	}
	/* check for sane offset value */
	if offset < 0 {
		log.Fatal("error: offset must be a positive integer value")
	}
	/* calculate base id */
	base := int(st.Sys().(*syscall.Stat_t).Uid)
	log.Println("calculated base", base)
	/* walk files and update uid and gid */
	if err = filepath.Walk(directory, walkFn(base, offset, verbose)); err != nil {
		log.Fatal("Error: ", err)
	}
}
